### PR TITLE
mathematica: Install desktop items

### PR DESCRIPTION
**Pull request by qbg**
###### Motivation for this change
Install the desktop items so users can launch Mathematica from their desktop environment, notebooks are associated with Mathematica, etc.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/science/math/mathematica/default.nix [hunk 1]
```diff
@@ -92,6 +92,10 @@ stdenv.mkDerivation rec {
     sed -i -e 's/^PATH=/# PATH=/' MathInstaller
     sed -i -e 's/\/bin\/bash/\/bin\/sh/' MathInstaller
 
+    # Install the desktop items
+    sed -i -e 's/Executables\/Mathematica/..\/..\/bin\/mathematica/' MathInstaller
+    export XDG_DATA_HOME="$out/share"
```
Comment (Artturin) [inline pkgs/applications/science/math/mathematica/default.nix:97]: Why is this required?

Comment (Artturin) [reply pkgs/applications/science/math/mathematica/default.nix:97]: Never mind i now know but can it be done in the install call like execdir is?

Comment (qbg) [reply pkgs/applications/science/math/mathematica/default.nix:97]: There doesn't appear to be any option in `MathInstaller` for configuring where these files go. `MathInstaller` calls the `xdg-desktop-menu` script (and the like) to install these files.

File: pkgs/applications/science/math/mathematica/default.nix [hunk 2]
```diff
@@ -92,6 +92,10 @@ stdenv.mkDerivation rec {
     sed -i -e 's/^PATH=/# PATH=/' MathInstaller
     sed -i -e 's/\/bin\/bash/\/bin\/sh/' MathInstaller
 
+    # Install the desktop items
+    sed -i -e 's/Executables\/Mathematica/..\/..\/bin\/mathematica/' MathInstaller
```
Comment (kanashimia) [inline pkgs/applications/science/math/mathematica/default.nix:96]:
```suggestion
    sed -i -e 's|Executables/Mathematica|../../bin/mathematica|' MathInstaller
```

Also can change line that is above like this too.

Comment (kanashimia) [reply pkgs/applications/science/math/mathematica/default.nix:96]: Also could even replace all of the `sed` calls to `substituteInPlace` as i see.

Comment (qbg) [reply pkgs/applications/science/math/mathematica/default.nix:96]: The regex for `PATH` would be tricky. I'll try updating the other two to use `substituteInPlace`.

Comment (qbg) [reply pkgs/applications/science/math/mathematica/default.nix]: I've replaced the two `sed` calls with `substituteInPlace`, and also expanded on the comment.

It still builds and works as before.

**Review by herberteuler — APPROVED** [review]
